paz/models/detection/model_util.py

```diff
@@ -0,0 +1,268 @@
+    Input:
+        point_cloud: TF tensor in shape (B,N,C)
+        mask: TF tensor in shape (B,N) of 0 (not pick) or 1 (pick)
+        npoints: int scalar, maximum number of points to keep (default: 512)
+    Output:
+        object_pc: TF tensor in shape (B,npoint,C)
+        indices: TF int tensor in shape (B,npoint,2)
+    """
+
+    def mask_to_indices(mask):
+        indices = np.zeros((mask.shape[0], npoints, 2), dtype=np.int32)
+        for i in range(mask.shape[0]):
+            pos_indices = np.where(mask[i, :] > 0.5)[0]
+            # skip cases when pos_indices is empty
+            if len(pos_indices) > 0:
+                if len(pos_indices) > npoints:
+                    choice = np.random.choice(len(pos_indices),
+                                              npoints, replace=False)
+                else:
+                    choice = np.random.choice(len(pos_indices),
+                                              npoints - len(pos_indices), replace=True)
+                    choice = np.concatenate((np.arange(len(pos_indices)), choice))
+                np.random.shuffle(choice)
+                indices[i, :, 1] = pos_indices[choice]
+            indices[i, :, 0] = i
+        return indices
+
+    indices = tf.py_function(mask_to_indices, [mask], tf.int32)
+    object_pc = tf.gather_nd(point_cloud, indices)
+    return object_pc, indices
+
+
+def parse_data(raw_record):
+    example = parse_frustum_point_record(raw_record)
+    return example['frustum_point_cloud'], \
+           tf.cast(example['one_hot_vec'], tf.float32), \
+           tf.cast(example['seg_label'], tf.int32), \
+           example['box3d_center'], \
+           tf.cast(example['angle_class'], tf.int32), \
+           example['angle_residual'], \
+           tf.cast(example['size_class'], tf.int32), \
+           example['size_residual']
+
+
+def parse_frustum_point_record(tfexample_message: str):
+    NUM_CLASS = 3
+    NUM_POINT = 1024
+    NUM_CHANNELS_OF_PC = 3
+
+    keys_to_features = {
+        "size_class": tf.io.FixedLenFeature((), tf.int64, tf.zeros((), tf.int64)),
+        "size_residual": tf.io.FixedLenFeature((3,), tf.float32, tf.zeros((3,), tf.float32)),
+        "seg_label": tf.io.FixedLenFeature((NUM_POINT,), tf.int64, tf.zeros((NUM_POINT,), tf.int64)),
+        "frustum_point_cloud": tf.io.FixedLenFeature((NUM_POINT, NUM_CHANNELS_OF_PC), tf.float32),
+        "rot_angle": tf.io.FixedLenFeature((), tf.float32, tf.zeros((), tf.float32)),
+        "angle_class": tf.io.FixedLenFeature((), tf.int64, tf.zeros((), tf.int64)),
+        "angle_residual": tf.io.FixedLenFeature((), tf.float32, tf.zeros((), tf.float32)),
+        "one_hot_vec": tf.io.FixedLenFeature((NUM_CLASS,), tf.int64),
+        "box3d_center": tf.io.FixedLenFeature((3,), tf.float32, tf.zeros((3,), tf.float32)),
+    }
+    parsed_example = tf.io.parse_single_example(tfexample_message, keys_to_features)
+    return parsed_example
+

```

Comment:
this should not be here. This is data specific thus it should in a data manager rather than in model creation.